`tctl sso`: update general SSO docs and GitHub guide. (#14227)
gravitational/teleport: docs/pages/enterprise/sso.mdx

</Tabs>

> Next, define an authentication connector. We recommend the usage of helper 

Reviewer: What would you think of adding step-by-step instructions for using a helper command to create the connector? We're already showing example snippets, so I think it would make sense to create a seamless step-by-step experience. We could use a `Tabs` component for this, with one `TabItem` per connector type.

Author: Hmm, not sure what that should look like? Do you mean to move individual steps in this guide into tabs? I'm probably missing the right idea here...

Reviewer: It would be something like this:

````markdown
Next, define an authentication connector using the `tctl sso configure` 
command:

<Tabs>
  <TabItem title="OIDC">

  Here is how to use `tctl sso configure oidc`.

  </TabItem>
  <TabItem title="SAML">

  Here is how to use `tctl sso configure saml`.

  </TabItem>
  <TabItem title="GitHub">

  Here is how to use `tctl sso configure gh`.

  </TabItem
</Tabs>
````
This way, the reader wouldn't need to find the relevant documentation on their own.

> commands to achieve this for [SAML](../setup/reference/cli.mdx#tctl-sso-configure-saml), 

Reviewer: If we add an example of the helper commands, we should remove the links here to prevent breaking the reader's flow.

</TabItem>
</Tabs>

> Note:

Reviewer: ```suggestion
```

> - Manually pasting the XML to `entity_descriptor` field is error prone. We recommend using `tctl sso configure saml -e entity_descriptor.xml ...` command instead.

Reviewer: ```suggestion
Manually pasting XML into the `entity_descriptor` field is error prone. We recommend using the `tctl sso configure saml -e entity_descriptor.xml ...` command instead.
```

- You may use `entity_descriptor_url`, in lieu of `entity_descriptor`, to fetch the entity descriptor from

>   your IDP. Though, we recommend "pinning" the entity descriptor by including the XML rather than fetching from a URL.

Reviewer: Does the helper command add the XML to pin the entity descriptor? If so, we should indicate this here, since that's the recommended approach.

> Test the connector with [`tctl sso test`](../setup/reference/cli.mdx#tctl-sso-test) command:

Reviewer: ```suggestion
Test the connector with the `tctl sso test` command:
```
so we don't break the reader's flow

Author: I'm not entirely convinced that including the link constitutes the breaking of reader focus: we are not forcing them to follow the link, but rather making an option to do so if they are interested. This is similar to inline-link style in [Wikipedia](https://en.wikipedia.org/wiki/Teleport_(open-source_software)#Architecture), for example:

> == Architecture ==
> 
> Teleport is written in [Go](Go (programming language)) programming language, and runs on [UNIX](Unix)-compatible operating systems, including Linux, [macOS](macOS), and several [BSD/OS](BSD) variants. Teleport consists of two [executable](executable)s: <code>tsh</code> (command line client) and <code>teleport</code> (server daemon).

Sure, the reader can get sidetracked by clicking all the links, but my expectation is that they only do it when they want to get some clarification and would otherwise be lost.

Let me know what you think, though.

Reviewer: I think with an inline link like this, the reader doesn't know what will be on the other side if they're not already familiar with the target URL. This means that a reader will need to open another browser window and shift context to make the determination of whether the link is useful. 

For a conceptual guide (e.g., a Wikipedia page), where a reader may already be browsing different sources of information, inline links are fine. For step-by-step guides, we're trying to keep all the relevant information on the same page. Otherwise, it can get frustrating to enter a tutorial thinking you'll achieve a goal by the end of it only to read incomplete instructions and run into roadblocks.

This guide is kind of a mix between a conceptual guide and a step-by-step guide. It has step-by-step instructions, but they're not comprehensive enough to follow as a tutorial. It covers a variety of topics conceptually, but not in much depth. The changes I'm suggesting make it more like a step-by-step guide.

Author: Ah, that makes sense, thank you for the explanation.

> ```bsh

Reviewer: ````suggestion
```code
````

## SSO Customization

| Provider  | YAML                 | Example                                               |
|-----------|----------------------|-------------------------------------------------------|

> | Github    | `display: Github`    | ![github](../../img/teleport-sso/github@2x.png)       |

Reviewer: "Github" should be "GitHub", since that's how the product is capitalized. Is `Github` enforced by `teleport`?

Author: No enforcement, just a typo.

If something is not working, we recommend to:

> - Run the connector definition through [`tctl sso test`](../setup/reference/cli.mdx#tctl-sso-test) command.

Reviewer: I'd suggest removing the link and adding an example of how to use `tctl sso test` to troubleshoot.

- Double-check the host names, tokens and TCP ports in a connector definition.

> - Double-check the role mapping in connector definition for possible typos.

Reviewer: ```suggestion
- Double-check the role mapping in your connector definition for possible typos.
```

> - Verify the configuration on IdP side (callback URLs, claim names, scopes, ...).

Reviewer: ```suggestion
- Verify the configuration on your IdP's side (e.g., callback URLs, claim names, and scopes).
```
